### ISSUE TITLE
systemd-firstboot: Schedule systemd-firstboot after dbus becomes available

### ISSUE THREAD

**Issue by frederictobiasc** [systemd/systemd#33637]
### Component

systemd-firstboot

### Is your feature request related to a problem? Please describe

I'm always frustrated when I write firstboot assistants and can't use dbus.
To launch my custom firstboot assistant, I use [`systemd-firstboot.service`](https://github.com/systemd/systemd/blob/main/units/systemd-firstboot.service) as a basis.

### Describe the solution you'd like

Schedule `systemd-firstboot.service` in such a way, that accessing dbus is possible without breaking the firstboot mechanism.

### Describe alternatives you've considered

Overriding the stock [`systemd-firstboot.service`](https://github.com/systemd/systemd/blob/main/units/systemd-firstboot.service) with 

```
Requires=dbus.service
```

leads to a dependency cycle:

```
# journalctl -u systemd-firstboot
Jul 05 09:09:42 m systemd[1]: sysinit.target: Found ordering cycle on systemd-firstboot.service/start
Jul 05 09:09:42 m systemd[1]: sysinit.target: Found dependency on dbus.service/start
Jul 05 09:09:42 m systemd[1]: sysinit.target: Found dependency on basic.target/start
Jul 05 09:09:42 m systemd[1]: sysinit.target: Found dependency on sysinit.target/start
Jul 05 09:09:42 m systemd[1]: sysinit.target: Job systemd-firstboot.service/start deleted to break ordering cycle starting with sysinit.target/start
```

### The systemd version you checked that didn't have the feature you are asking for

255

**Comment by poettering** [comment]
> Requires = ["dbus.service"];
> After = ["dbus.service"];

hmm? what is this? that's not systemd unit file syntax?

I am not sure I grok this request. The systemd-firstboot doesn't use D-Bus at all, so why would it have such a dependency?

If you want to write your own tool replacing systemd-firstboot tool, then you have to write your own unit file for it anyway, so you fix things there?

I am a bit confused by this request tbh?

**Comment by frederictobiasc** [comment]
Hi @poettering and thanks for your feedback!

>     Requires = ["dbus.service"];
>     After = ["dbus.service"];
>
> hmm? what is this? that's not systemd unit file syntax?

Right, please excuse me. That is Nix syntax, which gets translated into systemd unit files. I updated the original description.

> If you want to write your own tool replacing systemd-firstboot tool, then you have to write your own unit file for it anyway, so you fix things there?

I agree.

My perspective when creating the issue was: If I were to build such a firstboot replacement, I would try to stay as close as possible to the upstream solution to avoid breaking things in the future.

> The systemd-firstboot doesn't use D-Bus at all, so why would it have such a dependency?

To my understanding, `systemd-firstboot` uses a systemd internal D-Bus socket, since `/var/run/dbus/system_bus_socket` is not yet available at that time. For instance, [`reload_vconsole`](https://github.com/systemd/systemd/blob/14fb6354f18a4927213554af0adcff6441354505/src/firstboot/firstboot.c#L1618) uses D-Bus to restart `systemd-vconsole-setup.service`.

Since it is probably not a sensible approach for systemd-external applications to use interfaces that are intended exclusively for systemd-internal purposes, I would have to schedule a firstboot replacement at a time when D-Bus is available and would therefore have to deviate from the upstream solution.

However, as the boot process is already further advanced here, I am worried that this could lead to further problems.

To make life easier for the developers of a firstboot-replacement, I suggest changing the implementation of `systemd-firstboot` to the use of public interfaces to allow the developers of a firstboot-replacement to stay as close as possible to the upstream solution.

However, I would also understand if you think this is unreasonable. So please feel free to close this issue if you do.

**Comment by frederictobiasc** [comment]
> For instance, [reload_vconsole](https://github.com/systemd/systemd/blob/14fb6354f18a4927213554af0adcff6441354505/src/firstboot/firstboot.c#L1618) uses D-Bus to restart systemd-vconsole-setup.service.

Is this something we'll be able to do via varlink in the future?

**Comment by poettering** [comment]
> To my understanding, `systemd-firstboot` uses a systemd internal D-Bus socket, since `/var/run/dbus/system_bus_socket` is not yet available at that time. For instance, [`reload_vconsole`](https://github.com/systemd/systemd/blob/14fb6354f18a4927213554af0adcff6441354505/src/firstboot/firstboot.c#L1618) uses D-Bus to restart `systemd-vconsole-setup.service`.

Oh indeed, you are right, we do D-Bus.

**Comment by poettering** [comment]
So yeah, it leaves you indeed in kind of a shitty situation if you want to reimplement the logic, you either have to use our somewhat internal dbus socket, or do things in late boot.

**Comment by poettering** [comment]
> Is this something we'll be able to do via varlink in the future?

Yes, definitely. That would be the best solution for this I think. Once Varlink support is there we can support this from earliest boot on.
